tm4c123x-hal/src/eeprom.rs

```diff
@@ -0,0 +1,390 @@
+//! Code for the EEProm module.
+
+use core::convert::TryInto;
+
+use crate::sysctl::{self};
+use tm4c123x::EEPROM;
+pub use tm4c_hal::eeprom::{Blocks, Busy, EepromAddress, EepromError, Erase, Read, Write};
+
+// Number of EEPROM block on the TM4C123
+const EEPROM_BLOCK_SIZE: usize = 16;
+
+// Number of EEPROM blocks on the TM4C123
+const EEPROM_NUM_BLOCKS: usize = 32;
+
+// Total number of bytes in the EEPROM on the TM4C123
+const EEPROM_END_ADDRESS_BYTES: usize = 2048;
+
+// Total number of words in the EEPROM on the TM4C123
+const EEPROM_END_ADDRESS_WORDS: usize = 512;
+
+// Size of the EEPROM word in bytes
+const BYTES_PER_WORD: usize = 4;
+
+/// Eeprom struct
+pub struct Eeprom {
+    /// Eeprom registers
+    eeprom: EEPROM,
+}
+
+impl Eeprom {
+    /// Configures a new EEPROM struct using the datasheet section 8.2.4.2
+    pub fn new(eeprom: EEPROM, _pc: &sysctl::PowerControl) -> Self {
+        let final_eeprom = Eeprom { eeprom };
+
+        // See Section 8.2.4.2 EEPROM Initialization and Configuration
+        // in the datasheet:
+
+        // 0. Power on the EEPROM peripheral
+        sysctl::control_power(
+            _pc,
+            sysctl::Domain::Eeprom,
+            tm4c_hal::sysctl::RunMode::Run,
+            tm4c_hal::sysctl::PowerState::On,
+        );
+
+        // 1. The datasheet calls for at least a 6 cycle delay before polling
+        // the working register. Need to make sure the loop isn't optimized
+        // out.
+        final_eeprom.delay(20);
+
+        // 2. Poll busy
+        final_eeprom.wait();
+
+        // 3. Read PRETRY and ERETRY
+        if final_eeprom.eeprom.eesupp.read().eretry().bit_is_set() {
+            panic!("Eeprom ERETRY bit set, please investigate or stop using the EEPROM peripheral");
```

Comment:
Should this be a panic or should it return an error?

Reply:
I can change to an error, but chose a panic because of the datasheet (section 8.2.4.2):

If the PRETRY or ERETRY bits are set in the EESUPP register, the EEPROM was unable to recover its state. If power is stable when this occurs, this indicates a fatal error and is likely an indication that the EEPROM memory has exceeded its specified lifetime write/erase specification. If the supply voltage is unstable when this return code is observed, retrying the operation once the voltage is stabilized may clear the error.

Reply:
OK! Can you add a comment explaining this?

Reply:
Sorry, I had added code comments, but did you mean function comment? I have just added a comment to the Eeprom::new() functions that informs users that the function may panic, why, and with the quote from the datasheet explaining the issue.